### PR TITLE
Added Gaussian Blur module

### LINKED ISSUE #177: Add a Gaussian Blur module
Hi, this is a first-timers-only issue. This means we've worked to make it more legible to folks who either haven't contributed to our codebase before, or even folks who haven't contributed to open source before.

If that's you, we're interested in helping you take the first step and can answer questions and help you out as you do. Note that we're especially interested in contributions from people from groups underrepresented in free and open source software!

If you have contributed before, consider leaving this one for someone new, and looking through our general help wanted issues. Thanks!

The problem
Implementing a Gaussian blur module using https://www.npmjs.com/package/ndarray-gaussian-filter
We have already used this module in the edge detection module in this line  https://github.com/publiclab/image-sequencer/blob/5922b0de581f6ad3626c4f22713caeca870c336c/src/modules/EdgeDetect/Module.js#L28

Also have a look at this pr to submit a perfect module https://github.com/publiclab/image-sequencer/pull/161/files

Solution
Follow the patter of other modules and write a blur module see [contributing.md](https://github.com/publiclab/image-sequencer/blob/master/CONTRIBUTING.md) and do the actual image manipulation like in the above line

Steps to fix

 claim this issue with a comment
 set up a repository locally and make sure that all tests pass
 start a pull request (see contributing to Public Lab software) but mark it as "in progress"
 fix the issue following the steps above
 make sure all tests pass
 alert someone via either the developers list (see below) to have your pull request merged. This may take a few extra steps depending on when you do it, but we'll help you out!
Please email the developers list (see https://publiclab.org/developers) if you have questions!

### PR DESCRIPTION
**Pull request by danielleshwed**
Added Blur Module

Fixes #177

### CODE REVIEW

**Comment by jywarren** [comment]
Hi, wow, this looks super! @tech4GT @ccpandhare what do you think? 

I'm trying to follow, is the blurring actually connected up? Have you tested it out with an image, and could you post the before/after images?

Fantastic work! Thank you so much for your contribution!

**Comment by jywarren** [comment]
I added the issue number to the description and made the title a bit more descriptive too. Thanks!! 👍

**Comment by danielleshwed** [comment]
@jywarren thank you :) when i run the debug script I seem to not be getting any output images to the output folder, would you have any idea why this might be happening? (not just using blur also with other already implemented commands such as invert) 

Sorry this is my first Open Source contribution so learning as I go!

**Comment by tech4GT** [comment]
@danielleshwed oh, thanks a lot for detecting that, the debug script is not right, we can have a separate issue for that for now run this in your terminal, good work btw, please add screenshots as well😁
```
node ./index.js -i ./examples/images/monarch.png -s invert
```

File: src/modules/Blur/Module.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+         // Tell the UI that a step is being drawn
+         UI.onDraw(options.step);
+
+         var step = this;
+
+         function extraManipulation(pixels){
+           return pixels = require('ndarray-gaussian-filter')(pixels,options.blur)
+         }
+
+         function changePixel(r, g, b, a){
```
Comment (tech4GT) [inline src/modules/Blur/Module.js]: You don't have to use this function for this module, this function manipulates the brightness

Comment (tech4GT) [reply src/modules/Blur/Module.js]: pass a function which returns the current values of rgb

Comment (jywarren) [reply src/modules/Blur/Module.js]: so `extraManipulation()` could be renamed `blur()`, right? and then where would that be returned -- on what line? Thanks, @tech4GT !!

**Comment by danielleshwed** [comment]
@tech4GT i am still having issues testing using your suggested command... seems to run in the command line with no errors but no output images show up in the output folder

**Comment by tech4GT** [comment]
@danielleshwed that is because the of the empty output directory which was the original fto issue you asked to work on, ironical haha
Just delete the output directory and try again, it should work just fine😁

**Comment by danielleshwed** [comment]
Is this what we are looking for?

image1_0
image1_1

**Comment by tech4GT** [comment]
@danielleshwed this is not very right but i just checked and the module we are using is itself facing some issues so this output means that your code is right

**Comment by tech4GT** [comment]
@jywarren let's remove dependency on any external blur module altogether since i tried 2 more npm modules and none is free of caveats, we can implement this code ourselves, I will open a separate issue for that and start working on it right away what do you think?

**Comment by danielleshwed** [comment]
@tech4GT also just by the way, the only way I could get the issue with the output to be resolved is by adding the following code to ExportBin.js , simply deleting the output folder did not work :) 

```
  if(files.length === 0){
        cb(dirs);
        return [];
      }
```

**Comment by tech4GT** [comment]
@danielleshwed this issue is already fixed in pr #179 so that's taken care of.
Thanks anyway

**Comment by jywarren** [comment]
@danielleshwed i think @tech4GT is talking about the exportBin.js error in #179, so i hope that's no longer an issue with the latest code. 

No worries about the conflicts -- you can rebase these yourself if you'd like (there's [some guidance on this here](https://publiclab.org/wiki/developers#Resources)) or we can help you with that.

@tech4GT would you like to try to find a module that will work, or would you like to add commits to @danielleshwed's PR (either adding them yourself or opening a PR based on this branch) to solve the final blurring issue? 

@danielleshwed as @tech4GT said, it looks like your code is perfect and complete now, and we're just disappointed with the blur module we originally chose. Let's see if @tech4GT can help with the next step here. 

Thanks everybody! Great teamwork!

**Comment by tech4GT** [comment]
@jywarren I will add commits to the branch and i will generate a new pr coz I don't have rights to edit the pr😅

**Comment by tech4GT** [comment]
@jywarren this can be closed now please have a look at pr #186

**Comment by jywarren** [comment]
OK cool so #186 incorporates @danielleshwed 's commits, so if I merge that i'll combine all your collaborative work and merge it all in. Thanks!

**Comment by jywarren** [comment]
Solved in #186, thank you both so much!!! This was great.

**Comment by jywarren** [comment]
@danielleshwed I'm very sorry, i just checked and the "squashing" of commits has preserved your name in our commit log but not in the list of commiters. https://github.com/publiclab/image-sequencer/graphs/contributors

I'd like to properly attribute your work -- do you think you could open another PR even with a minor tweak, or alternatively if you're interested, we'd love to have your help with another feature request? Either way I want to ensure your work is recognized here; we're very grateful!

**Comment by danielleshwed** [comment]
@jywarren sure i'd like to work on another issue or feature :) Feel free to tag me in an issue which you need help on!

**Pull request closed without merging**